### PR TITLE
Hotfix exception handling

### PR DESCRIPTION
**Pull request by xolian**

### CODE REVIEW

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Below are required fields for qualifying your issue to our support team.
```
Comment (KylerWitting) [inline .github/ISSUE_TEMPLATE.md]: Here is a [link](https://github.com/angular-translate/angular-translate/blob/master/.github/ISSUE_TEMPLATE.md) for a more verbose issue template. Also I suggest adding a section to post logs.

File: .pytest_cache/v/cache/lastfailed [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (KylerWitting) [inline .pytest_cache/v/cache/lastfailed]: Remove

File: webbreaker/webinspectclient.py [hunk 3]
```diff
@@ -89,31 +90,34 @@ def create_scan(self):
...
-            Logger.app.critical("An Authorization Error occured.")
-            exit(1)
-        logger_response = json.dumps(response, default=lambda o: o.__dict__, sort_keys=True)
-        Logger.app.debug("Request sent to {0}:\n{1}".format(self.url, overrides))
-        Logger.app.debug("Response from {0}:\n{1}\n".format(self.url, logger_response))
+            api = webinspectapi.WebInspectApi(self.url, verify_ssl=False, username=self.username, password=self.password)
+            response = api.create_scan(overrides)
+            if response.response_code == 401:
+                Logger.app.critical("An Authorization Error occured.")
+                exit(1)
```
Comment (KylerWitting) [inline webbreaker/webinspectclient.py]: Change to new module

File: webbreaker/webinspectclient.py [hunk 4]
```diff
@@ -89,31 +90,34 @@ def create_scan(self):
...
-            Logger.app.error("No scan was launched!")
-            Logger.app.error("{}".format(response.message))
-            return False
-        return scan_id
+            if response.success:
+                scan_id = response.data['ScanId']
+                sys.stdout.write(str('WebInspect scan launched on {0} your scan id: {1}\n'.format(self.url, scan_id)))
+            else:
+                Logger.app.error("No scan was launched!\n {}".format(response.message))
+                exit(1)
```
Comment (KylerWitting) [inline webbreaker/webinspectclient.py]: Change to new module

File: webbreaker/webinspectclient.py [hunk 5]
```diff
@@ -89,31 +90,34 @@ def create_scan(self):
...
+            return scan_id
```
Comment (KylerWitting) [inline webbreaker/webinspectclient.py]: Move to if statement

File: webbreaker/webinspectqueryclient.py [hunk 6]
```diff
@@ -38,39 +38,45 @@ def export_scan_results(self, scan_id, scan_name, extension):
...
-        else:
-            Logger.app.critical("{}".format(response.message))
+        try:
+            api = webinspectapi.WebInspectApi(self.host, verify_ssl=False, username=self.username, password=self.password)
+            response = api.list_scans()
+            if response.success:
+                return response.data
+            elif response.response_code == 401:
+                Logger.app.critical("An Authorization Error occured.")
+                exit(1)
```
Comment (KylerWitting) [inline webbreaker/webinspectqueryclient.py]: Convert to new module

File: webbreaker/webinspectqueryclient.py [hunk 7]
```diff
@@ -38,39 +38,45 @@ def export_scan_results(self, scan_id, scan_name, extension):
...
+            if response.success:
+                try:
+                    with open('{0}.{1}'.format(scan_name, extension), 'wb') as f:
+                        Logger.app.info('Scan results file is available: {0}.{1}'.format(scan_name, extension))
+                        f.write(response.data)
+                except UnboundLocalError as e:
+                    Logger.app.error('Error saving file locally {}'.format(e))
+            elif response.response_code == 401:
+                Logger.app.critical("An Authorization Error occured.")
+                exit(1)
```
Comment (KylerWitting) [inline webbreaker/webinspectqueryclient.py]: Convert to new module

File: webbreaker/webinspectjitscheduler.py [hunk 8]
```diff
@@ -47,10 +53,13 @@ def __get_available_endpoints__(self):
         # introduce a random/arbitrary length delay before testing. Might help, might not.
         # Need a re-architecture or some kind of inter-process communication?
         possible_endpoints = self.__get_possible_endpoints__(max_concurrent_scans=self.max_scans)
+        # Added for debugging
```
Comment (KylerWitting) [inline webbreaker/webinspectjitscheduler.py]: Remove comment

File: webbreaker/webinspectjitscheduler.py [hunk 9]
```diff
@@ -17,6 +15,14 @@ def __init__(self, endpoints, size_list, size_needed='size_large', username=None
         self.password = password
         self.max_scans = self.__convert_size_to_count__()
 
+        # Added for debugging
```
Comment (KylerWitting) [inline webbreaker/webinspectjitscheduler.py]: Remove comment

File: webbreaker/__main__.py [hunk 10]
```diff
@@ -155,106 +161,120 @@ def webinspect(config):
               multiple=True,
               help="Assign workflow macro(s)")
 @pass_config
+@pass_config
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Remove duplication

File: webbreaker/__main__.py [hunk 11]
```diff
@@ -155,106 +161,120 @@ def webinspect(config):
...
     # ...and settings...
     try:
         webinspect_settings = webinspect_config.parse_webinspect_options(ops)
-    except (AttributeError, UnboundLocalError, IndexError):
-        Logger.app.error("Your configuration or settings are incorrect see log {}: ERROR: {}!!!".format(Logger.app_logfile))
-        exit(1)
+    except (AttributeError, UnboundLocalError, KeyError):
+        Logger.app.error(
+            "Your configuration or settings are incorrect see log: ERROR: {}!!!".format(Logger.app_logfile))
+
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Add exit failure code

File: webbreaker/__main__.py [hunk 12]
```diff
@@ -155,106 +161,120 @@ def webinspect(config):
...
     # OK, we're ready to actually do something now
 
     # The webinspect client is our point of interaction with the webinspect server farm
     try:
         webinspect_client = WebinspectClient(webinspect_settings, username=username, password=password)
     except (UnboundLocalError, EnvironmentError, NameError, TypeError) as e:
         Logger.app.critical("Incorrect WebInspect configurations found!! {}".format(str(e)))
-        exit(1)
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Convert to new module with failure exit code

File: webbreaker/__main__.py [hunk 13]
```diff
@@ -155,106 +161,120 @@ def webinspect(config):
...
-        # Change the provided policy name into the corresponding policy id for scan creation.
-        policy_id = webinspect_client.get_policy_by_guid(policy_guid)['id']
-        webinspect_client.scan_policy = policy_id
-        Logger.app.debug("New scan policy has been set")
+        else:
+            Logger.app.debug("No WebInspect Scan Override Policy was selected: {}!".format
+                             (webinspect_client.scan_policy))
 
+    except (UnboundLocalError, NameError) as e:
+        Logger.app.error("No WebInspect server was found: {}!".format(e))
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Failure exit code

File: webbreaker/__main__.py [hunk 14]
```diff
@@ -155,106 +161,120 @@ def webinspect(config):
...
-        webinspect_client.upload_webmacros()
+        # if there was a provided scan policy, we've already uploaded so don't bother doing it again. hack.
+        if webinspect_client.webinspect_upload_policy and not webinspect_client.scan_policy:
+            webinspect_client.upload_policy()
 
-    # if there was a provided scan policy, we've already uploaded so don't bother doing it again. hack.
-    if webinspect_client.webinspect_upload_policy and not webinspect_client.scan_policy:
-        webinspect_client.upload_policy()
+    except (UnboundLocalError, NameError) as e:
+        Logger.app.error("No WebInspect server was found: {}!".format(e))
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Insert exit code

File: webbreaker/__main__.py [hunk 15]
```diff
@@ -273,25 +293,32 @@ def scan(config, **kwargs):
 
             if status.lower() != 'complete':  # case insensitive comparison is tricky. this should be good enough for now
                 Logger.app.error("See the WebInspect server scan log --> {}, typically the application to be scanned is"
-                     "unavailable.".format(WebInspectConfig().endpoints))
-                #Logger.app.error('Scan is incomplete and is unrecoverable. WebBreaker will exit!!')
+                                 "unavailable.".format(WebInspectConfig().endpoints))
+                # Logger.app.error('Scan is incomplete and is unrecoverable. WebBreaker will exit!!')
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Remove comment

File: webbreaker/__main__.py [hunk 16]
```diff
@@ -273,25 +293,32 @@ def scan(config, **kwargs):
...
+                sys.exit(ExitStatus.failure)
+
         webinspect_client.export_scan_results(scan_id, 'fpr')
         webinspect_client.export_scan_results(scan_id, 'xml')
 
-    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError, NameError) as e:
+    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError, NameError, KeyError, IndexError) as e:
         Logger.app.error(
-            "Unable to connect to WebInspect {0}, see also: {1}".format(webinspect_settings['webinspect_url'], e))
+            #"Unable to connect to WebInspect {0}, see also: {1}".format(webinspect_settings['webinspect_url'], e))
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Remove comment

File: webbreaker/__main__.py [hunk 17]
```diff
@@ -273,25 +293,32 @@ def scan(config, **kwargs):
...
-    handle_scan_event('scan_end')
-    Logger.app.info("Webbreaker WebInspect has completed.")
+
+    try:
+        handle_scan_event('scan_end')
+        Logger.app.info("Webbreaker WebInspect has completed.")
+
+    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError, TypeError, NameError, KeyError, IndexError) as e:
+        Logger.app.error(
+            #"Unable to connect to WebInspect {0}, see also: {1}".format(webinspect_settings['webinspect_url'], e))
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Remove comment

File: webbreaker/__main__.py [hunk 18]
```diff
@@ -497,117 +524,109 @@ def download(config, server, scan_name, scan_id, x, protocol, username, password
...
-        auth_config.write_username(username)
-        auth_config.write_password(password)
+        # If we've made it this far, our new credentials are valid and should be saved
+        if username is not None and password is not None and not auth_config.has_auth_creds():
+            auth_config.write_username(username)
+            auth_config.write_password(password)
+
+    except (UnboundLocalError, EnvironmentError) as e:
+        Logger.app.critical("Incorrect WebInspect configurations found!! {}".format(e))
+        exit(1)
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Change to sys.exit(ExitStatus.failure)

File: webbreaker/__main__.py [hunk 19]
```diff
@@ -666,7 +685,7 @@ def fortify_list(config, fortify_user, fortify_password, application):
         Logger.app.info("Fortify list has successfully completed")
     except ValueError:
         Logger.app.error("Unable to obtain a Fortify API token. Invalid Credentials")
-    except (AttributeError, UnboundLocalError) as e:
+    except (AttributeError, UnboundLocalError, TypeError) as e:
         Logger.app.critical("Unable to complete command 'fortify list': {}".format(e))
```
Comment (KylerWitting) [inline webbreaker/__main__.py]: Change to sys.exit(ExitStatus.failure)

File: .github/ISSUE_TEMPLATE.md (hunk 1 above)
Comment (xolian) [reply .github/ISSUE_TEMPLATE.md]: matt modified content per advice

**Comment by xolian** [comment]
modified documentation post review, merging to master

**Pull request merged**